src/pharmpy/modeling/estimation_steps.py

```diff
@@ -402,11 +404,147 @@ def set_evaluation_step(model: Model, idx: int = -1):
     return model.update_source()
 
 
+def add_derivative(model: Model, with_respect_to: Optional[Union[tuple[str], str]] = None):
+    """
+    Add a derivative to be calculcated when running the model. Currently, only
+    derivatives with respect to the prediction is supported.
+    First order derivates are specied either by single string or single-element tuple.
+    For instance with_respect_to = "ETA_1" or with_respect_to = ("ETA_1",)
+
+    Second order derivatives are specified by giving the two independent varibles in a tuple
+    of tuples. For instance with_respect_to ((ETA_1, EPS_1),)
+
+    Currently, only ETAs and EPSILONs are supported
+
+    Parameters
+    ----------
+    model : Model
+        Pharmpy modeas.
+    with_respect_to : Union[tuple, str]
+        Parameter name(s) to use as independent variables
+
+    Returns
+    -------
+    Pharmpy model.
+
+    """
+
+    if with_respect_to is None:
+        with_respect_to = tuple(
+            subset_combinations(
+                model.random_variables.etas.names, model.random_variables.epsilons.names
+            )
+        )
+
+    elif isinstance(with_respect_to, str):
+        with_respect_to = ((Expr.symbol(with_respect_to),),)
+    elif isinstance(with_respect_to, tuple):
+        with_respect_to = (tuple(Expr.symbol(w) for w in with_respect_to),)
+    else:
+        raise ValueError(
+            f"with_respect_to argument need to be tuple, str or None. Recieved {type(with_respect_to)}"
+        )
+
+    for derivative in with_respect_to:
+        for parameter in derivative:
+            if parameter not in model.random_variables.symbols:
+                raise ValueError(f"Parameter '{parameter}' not part of ETAs or EPS")
+
+    steps = model.execution_steps
+    new_step = steps[-1]
+
+    for new_derivative in EstimationStep._canonicalize_derivatives(with_respect_to):
```

Comment:
I dont think the replace method would solve this issue. The problem lies in that the order of second order derivatives is also canonicalized, and might differ from what the user specified. For example, if we have stored the derivatives `((ETA_1, ETA_2), ETA_1)` and we wish to remove `(ETA_2, ETA_1)` from that list, we need to be able to compare the two.
Instead of using the private `_canonicalize_derivates` method, this can be done using sets.